src/frontends/onnx/frontend/src/op/scan.cpp

```diff
@@ -0,0 +1,144 @@
+// Copyright (C) 2018-2022 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "op/scan.hpp"
+
+#include <iterator>
+#include <memory>
+
+#include "core/graph.hpp"
+#include "default_opset.hpp"
+#include "exceptions.hpp"
+#include "ngraph/function.hpp"
+#include "ngraph/log.hpp"
+#include "ngraph/op/util/op_types.hpp"
+#include "onnx_import/core/null_node.hpp"
+#include "utils/reshape.hpp"
+
+namespace ngraph {
+namespace onnx_import {
+namespace op {
+
+namespace {
+OutputVector generate_scan(const Node& node,
+                           int64_t default_axis,
+                           int64_t in_offset,
+                           std::string&& in_directions_attr_name) {
+    const auto& ng_inputs = node.get_ng_inputs();
+
+    const auto& subgraphs = node.get_subgraphs();
+    auto body_graph = subgraphs.at("body");
+    auto body_outputs = body_graph->get_ng_outputs();
+    auto body_inputs = body_graph->get_ng_parameters();
+
+    const int64_t num_scan_inputs = node.get_attribute_value<int64_t>("num_scan_inputs");
+    const size_t num_initial_values = body_inputs.size() - num_scan_inputs;
+    const size_t num_scan_outputs = body_outputs.size() - num_initial_values;
+
+    std::vector<int64_t> scan_input_axes =
+        node.get_attribute_value<std::vector<int64_t>>("scan_input_axes",
+                                                       std::vector<int64_t>(num_scan_inputs, default_axis));
+    std::vector<int64_t> scan_input_directions =
+        node.get_attribute_value<std::vector<int64_t>>(in_directions_attr_name,
+                                                       std::vector<int64_t>(num_scan_inputs, 0));
+    std::vector<int64_t> scan_output_axes =
+        node.get_attribute_value<std::vector<int64_t>>("scan_output_axes",
+                                                       std::vector<int64_t>(num_scan_outputs, default_axis));
+    std::vector<int64_t> scan_output_directions =
+        node.get_attribute_value<std::vector<int64_t>>("scan_output_directions",
+                                                       std::vector<int64_t>(num_scan_outputs, 0));
+
+    // Body inputs alignment
+    for (size_t i = 0; i < num_initial_values; ++i) {
+        body_inputs[i]->set_element_type(ng_inputs[i + in_offset].get_element_type());
+        body_inputs[i]->set_partial_shape(ng_inputs[i + in_offset].get_partial_shape());
+        body_inputs[i]->validate_and_infer_types();
+    }
```

Comment:
Why is it required?

Reply:
Actually I don't think it is required, current tests are working without this piece of code, but it helps to ensure shape and type compatibility between init values and the state params.
@mateusztabaka What do you think?

Reply:
TypeProto for inputs in subgraphs is optional so we have this loop to infer shape and element type from corresponding Scan inputs. Otherwise, body_inputs (which are Parameters) may have e.g. element::undefined type.